types/src/transaction.rs

```diff
@@ -1223,62 +1104,98 @@ impl Transaction {
+    where
+        S: ser::Serializer,
+    {
+        use ser::SerializeTuple;
 
-impl CanonicalSerialize for Transaction {
-    fn serialize(&self, serializer: &mut impl CanonicalSerializer) -> Result<()> {
         match self {
             Transaction::UserTransaction(txn) => {
-                serializer.encode_u8(TransactionVariantTag::UserTransaction.into())?;
-                serializer.encode_struct(txn)?;
+                let mut t = serializer.serialize_tuple(2)?;
+                t.serialize_element(&0u8)?;
+                t.serialize_element(txn)?;
+                t.end()
             }
             Transaction::WriteSet(write_set) => {
-                serializer.encode_u8(TransactionVariantTag::WriteSet.into())?;
-                serializer.encode_struct(write_set)?;
+                let mut t = serializer.serialize_tuple(2)?;
+                t.serialize_element(&1u8)?;
+                t.serialize_element(write_set)?;
+                t.end()
             }
             Transaction::BlockMetadata(block_metadata) => {
-                serializer.encode_u8(TransactionVariantTag::BlockMetadata.into())?;
-                serializer.encode_struct(block_metadata)?;
+                let mut t = serializer.serialize_tuple(2)?;
+                t.serialize_element(&2u8)?;
+                t.serialize_element(block_metadata)?;
+                t.end()
             }
-        };
-
-        Ok(())
+        }
     }
 }
 
-impl CanonicalDeserialize for Transaction {
-    fn deserialize(deserializer: &mut impl CanonicalDeserializer) -> Result<Self> {
-        let tag = TransactionVariantTag::try_from(deserializer.decode_u8()?)?;
-        let transaction = match tag {
-            TransactionVariantTag::UserTransaction => {
-                Transaction::UserTransaction(deserializer.decode_struct()?)
+impl<'de> de::Deserialize<'de> for Transaction {
+    fn deserialize<D>(deserializer: D) -> std::result::Result<Self, D::Error>
+    where
+        D: de::Deserializer<'de>,
+    {
+        struct TransactionVisitor;
+        impl<'de> de::Visitor<'de> for TransactionVisitor {
+            type Value = Transaction;
+
+            fn expecting(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+                formatter.write_str("Transaction")
             }
-            TransactionVariantTag::WriteSet => Transaction::WriteSet(deserializer.decode_struct()?),
-            TransactionVariantTag::BlockMetadata => {
-                Transaction::BlockMetadata(deserializer.decode_struct()?)
+
+            fn visit_seq<A>(self, mut seq: A) -> std::result::Result<Self::Value, A::Error>
+            where
+                A: de::SeqAccess<'de>,
+            {
+                use serde::de::Error;
+
+                let variant: u8 = seq
+                    .next_element()?
+                    .ok_or_else(|| A::Error::custom("expected u8"))?;
+                match variant {
+                    0 => {
```

Comment:
Prefer that everything is typed. What happened to TransactionVariantTag?

Reply:
This only exists in this form because the old implementation for this type was done incorrectly as the variant index was encoded using a u8 instead of a u32. Once this is fixed (after this change is landed) we can completely remove this section of code and just use the derive implementation.

Reply:
"just use the derive implementation"

^ does it mean if we change the order of variants in an enum, we break the lcs form of it?

Reply:
It'll be nice if we require explicitly tagging the variants with u32 numbers.